### PR TITLE
Fix application uplink queue message handling

### LINKED ISSUE #3653: Flaky Network Server tests
#### Summary

Running the Network Server unit tests fails.

#### Steps to Reproduce

1. `git checkout v3.11`
2. `docker run --rm -it -p 6379:6379 redis:5` 
3. `go test ./pkg/networkserver -failfast`

#### What do you see now?

```
$ go test ./pkg/networkserver/redis/ -failfast
--- FAIL: TestDownlinkTaskQueue (0.01s)
    --- FAIL: TestDownlinkTaskQueue/1st_run (0.00s)
        downlink_task_queue.go:95: Pop returned on empty schedule with error: 
            -> error:pkg/redis:store (store error) (attributes: map[])
            --> ERR Error running script (call to f_32bd83583b1367d88c7534ddb33d6622a3c58cc7): @user_script:8: @user_script: 8: This Redis command is not allowed from scripts  (attributes: map[])
    downlink_task_queue_test.go:32: Skipping 2nd run
--- FAIL: TestApplicationUplinkQueue (0.01s)
    --- FAIL: TestApplicationUplinkQueue/1st_run (0.00s)
        application_uplink_queue.go:198: Received unexpected Pop error: error:pkg/redis:store (store error)
        doc.go:66: 
            /home/ubuntu/lorawan-stack/pkg/networkserver/internal/test/shared/application_uplink_queue.go:198
            Expected: true
            Actual:   false
    application_uplink_queue_test.go:32: Skipping 2nd run
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 995ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
FAIL
FAIL	go.thethings.network/lorawan-stack/v3/pkg/networkserver/redis	0.270s
FAIL

```

CI also seems to fail randomly on a different test (see failure on https://github.com/TheThingsNetwork/lorawan-stack/runs/1684533672?check_suite_focus=true, but https://github.com/TheThingsNetwork/lorawan-stack/actions/runs/478857664 passes)

#### What do you want to see instead?

Tests should pass

#### Environment

- TTS 3.11
- Redis 5 (I am also getting errors locally with Redis 6, but CI seems to work just fine).

#### Can you do this yourself and submit a Pull Request?

No

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

Fix application uplink queue message ACK handling 
Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/3653

#### Changes

- Reimplement application uplink queue message handling
- Send messages in order - first join accepts, then invalidations, then everything else
- Recover messages, which failed to send

#### Testing

Unit tests

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Review by adriansmares — APPROVED** [review]

**Review by johanstokking — APPROVED** [review]

**Comment by rvolosatovs** [comment]
@adriansmares the fixes for invalidations are in-progress, but I'll wait for the API changes first and will do it in another PR.

**Comment by rvolosatovs** [comment]
`XAUTOCLAIM` implementation will be redundant after https://github.com/TheThingsNetwork/lorawan-stack/issues/3592 is done

File: CHANGELOG.md [hunk 1]
```diff
@@ -28,9 +28,7 @@ For details about compatibility between different releases, see the **Commitment
 - Ocassional race condition in uplink matching with replicated Network Server instances.
 - Ocassional race condition when matching pending sessions.
 - Conflict error when registering an end device via the wizard in the Console.
-- Pagination in the `List` and `ListTokens` RPCs of the `OAuthAuthorizationRegistry`.
-- Event name on user login.
-- Retrying downlink messages with absolute time on other gateways when one downlink path reports a scheduling conflict.
```
Comment (htdvisser) [inline CHANGELOG.md]: What's up with this?

**Review by adriansmares — COMMENTED** [review]
Implementation LGTM 👍 (`lua`/`go`). The changelog should indeed get fixed.

This PR references https://github.com/TheThingsNetwork/lorawan-stack/issues/44 indirectly, as it adds `XCLAIM` / `XPENDING` support.

**Review by johanstokking — APPROVED** [review]
LGTM except changelog

**Comment by rvolosatovs** [comment]
@adriansmares I'm seeing
```
 DEBUG Enqueue application uplinks for sending to Application Server device_uid=test-app2.test-dev-c namespace=networkserver started_at=2021-03-02 15:20:36.318452796 +0000 UTC uplink_count=1
 DEBUG Publish events                           device_uid=test-app2.test-dev-c event_count=3 namespace=networkserver started_at=2021-03-02 15:20:36.318452796 +0000 UTC
  WARN Failed to process upstream message       device_uid=test-app2.test-dev-c error=error:pkg/applicationserver:no_device_session (no device session; check device activation) grpc.method=HandleUplink grpc.service=ttn.lorawan.v3.NsAs namespace=grpc request_id=01EZSRDNQ38JY776J42RPPHMEF
```
in the log and on first try device ended up in a state where it was impossible to schedule downlinks anymore due to FCnt used by AS being too low. I'm merging this assuming this will be fixed in https://github.com/TheThingsNetwork/lorawan-stack/issues/3649

**Pull request merged**